README.md
# vscpl1drv-socketcan

Level I driver for Socketcan.


Copyright (C) 2000-2020 Ake Hedman, Grodans Paradis AB - MIT license.


debian_orig/changelog
vscpl1drv-socketcan (%MAJOR-VERSION.%MINOR-VERSION.%RELEASE-VERSION-%RELEASE-DEBIAN) stable; urgency=medium

  * Initial release

 -- Ake Hedman <akhe@grodansparadis.com>  %DATENOW


linux/test.cpp
// test.cpp :
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version
// 2 of the License, or (at your option) any later version.
//
// This file is part of the VSCP (http://www.vscp.org)
//
// Copyright (C) 2000-2020 Ake Hedman,
// Ake Hedman, Grodans Paradis AB, <akhe@grodansparadis.com>
//
// This file is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this file see the file COPYING.  If not, write to
// the Free Software Foundation, 59 Temple Place - Suite 330,
// Boston, MA 02111-1307, USA.
//
// $RCSfile: test.cpp,v $
// $Date: 2005/09/08 07:56:00 $
// $Author: akhe $
// $Revision: 1.3 $

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <dlfcn.h>

#include <canal.h>
#include <canal_macro.h>
#include <canaldlldef.h>

#define NUMBER_OF_THREADS 2

// Prototypes
void *workThread( void *id );

int main( void )
{
	int i;
	pthread_t workthreads[ NUMBER_OF_THREADS ];
	pthread_attr_t thread_attr;

	pthread_attr_init( &thread_attr );

	printf("\n\n\n");
	for ( i=0; i<NUMBER_OF_THREADS; i++ ) {

		// Create the worker thread.
		if ( pthread_create( 	&workthreads[ i ],
									&thread_attr,
									workThread,
									&i ) ) {
		}

	}

	// Give the threads some time to work
	printf("Giving threads some time to work.\n");
	sleep( 5 );

	for ( i=0; i<NUMBER_OF_THREADS; i++ ) {
		int *rv;
		pthread_join( workthreads[ i ], (void **)&rv );
	}

	return ( EXIT_SUCCESS );
}

///////////////////////////////////////////////////////////////////////////////
// workThread
//

void *workThread( void *id )
{
	int rv = 0;
	void* handle = 0;
	//parent *(*open)(void) = 0
	//const char* error = 0;
	LPFNDLL_CANALOPEN			        m_proc_CanalOpen;
	LPFNDLL_CANALCLOSE				    m_proc_CanalClose;
	LPFNDLL_CANALGETLEVEL			    m_proc_CanalGetLevel;
	LPFNDLL_CANALSEND				    m_proc_CanalSend;
	LPFNDLL_CANALRECEIVE			    m_proc_CanalReceive;
	LPFNDLL_CANALDATAAVAILABLE		    m_proc_CanalDataAvailable;
	LPFNDLL_CANALGETSTATUS			    m_proc_CanalGetStatus;
	LPFNDLL_CANALGETSTATISTICS		    m_proc_CanalGetStatistics;
	LPFNDLL_CANALSETFILTER			    m_proc_CanalSetFilter;
	LPFNDLL_CANALSETMASK			    m_proc_CanalSetMask;
	LPFNDLL_CANALSETBAUDRATE		    m_proc_CanalSetBaudrate;
	LPFNDLL_CANALGETVERSION			    m_proc_CanalGetVersion;
	LPFNDLL_CANALGETDLLVERSION		    m_proc_CanalGetDllVersion;
	LPFNDLL_CANALGETVENDORSTRING	    m_proc_CanalGetVendorString;
	// Generation 2
	LPFNDLL_CANALBLOCKINGSEND		    m_proc_CanalBlockingSend;
	LPFNDLL_CANALBLOCKINGRECEIVE	    m_proc_CanalBlockingReceive;
	LPFNDLL_CANALGETDRIVERINFO		    m_proc_CanalGetdriverInfo;

	handle = dlopen("./can232drv.so", RTLD_LAZY);

	if ( 0 == handle ) {
		fprintf( stderr, "test: Failed to load library %s\n", dlerror() );
		exit ( EXIT_FAILURE );
	}

	// * * * * CANAL OPEN * * * *
	if ( 0 == ( m_proc_CanalOpen = (LPFNDLL_CANALOPEN)dlsym( handle, "CanalOpen" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalOpen. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL CLOSE * * * *
	if ( 0 == ( m_proc_CanalClose = (LPFNDLL_CANALCLOSE)dlsym( handle, "CanalClose" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalClose. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL GETLEVEL * * * *
	if ( 0 == ( m_proc_CanalGetLevel = (LPFNDLL_CANALGETLEVEL)dlsym( handle, "CanalGetLevel" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalGetLevel. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL SEND * * * *
	if ( 0 == ( m_proc_CanalSend = (LPFNDLL_CANALSEND)dlsym( handle, "CanalSend" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalSend. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}


	// * * * * CANAL DATA AVAILABLE * * * *
	if ( 0 == ( m_proc_CanalDataAvailable = (LPFNDLL_CANALDATAAVAILABLE)dlsym( handle, "CanalDataAvailable" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalDataAvailable. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL DATA AVAILABLE * * * *
	if ( 0 == ( m_proc_CanalDataAvailable = (LPFNDLL_CANALDATAAVAILABLE)dlsym( handle, "CanalDataAvailable" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalDataAvailable. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL RECEIVE * * * *
	if ( 0 == ( m_proc_CanalReceive = (LPFNDLL_CANALRECEIVE)dlsym( handle, "CanalReceive" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalReceive. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL GET STATUS * * * *
	if ( 0 == ( m_proc_CanalGetStatus = (LPFNDLL_CANALGETSTATUS)dlsym( handle, "CanalGetStatus" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalGetStatus. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL GET STATISTICS * * * *
	if ( 0 == ( m_proc_CanalGetStatistics = (LPFNDLL_CANALGETSTATISTICS)dlsym( handle, "CanalGetStatistics" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalGetStatistics. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL SET FILTER * * * *
	if ( 0 == ( m_proc_CanalSetFilter = (LPFNDLL_CANALSETFILTER)dlsym( handle, "CanalSetFilter" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalSetFilter. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL SET MASK * * * *
	if ( 0 == ( m_proc_CanalSetMask = (LPFNDLL_CANALSETMASK)dlsym( handle, "CanalSetMask" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalSetMask. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL GET VERSION * * * *
	if ( 0 == ( m_proc_CanalGetVersion = (LPFNDLL_CANALGETVERSION)dlsym( handle, "CanalGetVersion" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalGetVersion. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL GET DLL VERSION * * * *
	if ( 0 == ( m_proc_CanalGetDllVersion = (LPFNDLL_CANALGETDLLVERSION)dlsym( handle, "CanalGetDllVersion" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalGetDllVersion. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	// * * * * CANAL GET VENDOR STRING * * * *
	if ( 0 == ( m_proc_CanalGetVendorString = (LPFNDLL_CANALGETVENDORSTRING)dlsym( handle, "CanalGetVendorString" ) ) )  {

		// Free the library
		dlclose( handle );
		fprintf( stderr, "test: Unable to get dl entry for CanalGetVendorString. error=%s", dlerror() );
		exit ( EXIT_FAILURE );

	}

	////////////////////////////////////////////////////////////////////////////
	//                           * * * T E S T S * * *
	///////////////////////////////////////////////////////////////////////////

	// Open
	char device[ 32 ];
	char flags[ 32 ];
	long dh;
	sprintf( device, "/dev/ttyS0;19200;0;0;125" );
	dh = (*m_proc_CanalOpen)( device, 1 );

	printf("Sending data.\n");

	canalMsg msg;

	//for ( int i= 0; i<256; i++ ) {
	for ( int i= 0; i<10; i++ ) {

		msg.obid = 1946;
		msg.flags = CANAL_IDFLAG_EXTENDED;
		msg.id = i;
		msg.sizeData = 8;
		msg.data[0] = 0x11;
		msg.data[1] = 0x22;
		msg.data[2] = 0x33;
		msg.data[3] = 0x44;
		msg.data[4] = 0x55;
		msg.data[5] = 0x66;
		msg.data[6] = 0x77;
		msg.data[7] = 0x88;
		msg.timestamp = 3210;
		(*m_proc_CanalSend)( dh, &msg );
		printf(".");

		msg.obid = 2001;
		msg.flags = CANAL_IDFLAG_EXTENDED;
		msg.id = i;
		msg.sizeData = 4;
		msg.data[0] = 0xA1;
		msg.data[1] = 0xB2;
		msg.data[2] = 0xC3;
		msg.data[3] = 0xD4;
		msg.data[4] = 0xE5;
		msg.data[5] = 0xF6;
		msg.data[6] = 0x07;
		msg.data[7] = 0x18;
		msg.timestamp = 0xcc55;
		(*m_proc_CanalSend)( dh, &msg );
		printf(".");
	}

	printf("\n");

	printf("===========================================================\n");
	printf("=                       Data for DLL\n");
	printf("===========================================================\n");

	printf( "Canal Level = %lu\n", (*m_proc_CanalGetLevel)( dh ) );
	printf( "Canal Version = %lu\n", (*m_proc_CanalGetVersion)() );
	printf( "DLL Version = %lu\n", (*m_proc_CanalGetDllVersion)() );
	printf( "Vendor String = %s\n", (*m_proc_CanalGetVendorString)() );

	printf("===========================================================\n");

	printf("Closing Log function\n");

	// Give the driver some time to finish writing
	printf("Waiting before close\n\n\n");
	sleep( 5 );

	// Close
	(*m_proc_CanalClose)( dh );

	printf("Closing DLL\n");

	dlclose( handle );

	pthread_exit( &rv );
}


linux/socketcanobj.h
///////////////////////////////////////////////////////////////////////////////
// CAN232Obj.h: interface for the CSocketcanObj class.
//
// This file is part is part of CANAL (CAN Abstraction Layer)
// http://www.vscp.org)
//
// Copyright (C) 2000-2020 Ake Hedman,
// Ake Hedman, Grodans Paradis AB, <akhe@grodansparadis.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//

#if !defined(SOCKETCANDRV_H__16828641_5EDF_4115_9522_97BD178F566B__INCLUDED_)
#define SOCKETCANDRV_H__16828641_5EDF_4115_9522_97BD178F566B__INCLUDED_

#if _MSC_VER > 1000
#pragma once
#endif // _MSC_VER > 1000


#define _POSIX
#include <stdio.h>
#include "unistd.h"
#include "stdlib.h"
#include <string.h>
#include "limits.h"
#include "syslog.h"
#include <net/if.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <pthread.h>

// Different on Kernel 2.6 and cansocket examples
// currently using locally from can-utils
// TODO remove include form makefile when they are in sync
#include <linux/can.h>
#include <linux/can/raw.h>

#include <canal.h>
#include <vscp.h>
#include <canal_macro.h>
//#include <com.h>
#include <dllist.h>


#define SOCKETCAN_BUF_SIZE		80	// Size for one message buffer

#define SOCKETCAN_MAX_RCVMSG	512	// Maximum number of received messages
#define SOCKETCAN_MAX_SNDMSG	512	// Maximum number of received messages

#define SOCKETCAN_RX_MUTEX     "___SOCKETCAN_LEVEL1_RX_MUTEX___"
#define SOCKETCAN_TX_MUTEX     "___SOCKETCAN_LEVEL1_TX_MUTEX___"

////////////////////////////////////////////////////////////////////////////////////////
// _socketcanobj
//

struct _socketcanobj {
    /*!
        Work flag for thread
     */
    bool m_bRun;

    /*!
     *  Device name
     */
    char m_devname[IFNAMSIZ + 1];

    /*!
     *  Socket to socketcan
     */
    int m_sock;

    /*!
        Send/Receive Statistics
     */
    canalStatistics m_stat;

    DoubleLinkedList m_rcvList; // List for incoming messages
    DoubleLinkedList m_sndList; // List for outgoing messages
};

class CSocketcanObj {
public:
    // Constructor
    CSocketcanObj();

    // Destructor
    virtual ~CSocketcanObj();

    /*!
        Open the device
        Return True on success
     */
    bool open(const char *pDevice, unsigned long flags);

    /*!
        Close the device
     */
    int close(void);

    /*!
        Send frame
     */
    int writeMsg( bool bExtended,
					unsigned long id,
					unsigned char dlc,
					unsigned char * pdata);

    /*!
        Send frame
     */
    int writeMsg(PCANALMSG pCanalMsg);


    /*!
        Receive frame
     */
    int readMsg(canalMsg *pMsg);


    /*!
        Set filter(code) and mask
     */
    bool setFilter(unsigned long filter, unsigned long mask);

    /*!
        Set filter (acceptance code)
     */
    bool setFilter(unsigned long filter);


    /*!
        Set mask
     */
    bool setMask(unsigned long mask);

    /*!
        Get statistics
     */
    bool getStatistics(PCANALSTATISTICS& pCanalStatistics);


    /*
        Get number of frames availabel in the input queue
     */
    int dataAvailable(void);

    /*!
        Get the status code
     */
    bool getStatus(PCANALSTATUS pCanalStatus);


    struct _socketcanobj m_socketcanobj;


    /*!
        The socketcan read/write mutexes
	 */
    pthread_mutex_t m_socketcanRcvMutex;
	pthread_mutex_t m_socketcanSndMutex;

    /*!
        id for worker thread
     */
    pthread_t m_threadId;

};

#endif // !defined(SOCKETCANDRV_H__16828641_5EDF_4115_9522_97BD178F566B__INCLUDED_)


linux/socketcanobj.cpp
///////////////////////////////////////////////////////////////////////////////
// socketcanobj.cpp: implementation of the CSocketcanObj class.
//
// This file is part is part of CANAL (CAN Abstraction Layer)
// http://www.vscp.org)
//
// Copyright (C) 2000-2020 Ake Hedman,
// Ake Hedman, Grodans Paradis AB, <akhe@grodansparadis.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//

#include <stdio.h>
#include "unistd.h"
#include "stdlib.h"
#include <string.h>
#include "limits.h"
#include "syslog.h"
#include <net/if.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>

// Different on Kernel 2.6 and socketcan examples
// currently using locally from can-utils
// TODO remove include form makefile when they are in sync
#include <linux/can.h>
#include <linux/can/raw.h>

#include <signal.h>
#include <ctype.h>
#include <libgen.h>
#include <time.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/uio.h>
#include <net/if.h>
#include "socketcanobj.h"

// Prototypes
void *workThread(void *p);
bool socketcanToCanal(char * p, PCANALMSG pMsg);

////////////////////////////////////////////////////////////////////////////////
// Construction/Destruction
////////////////////////////////////////////////////////////////////////////////

CSocketcanObj::CSocketcanObj()
{
    strcpy(m_socketcanobj.m_devname, "vcan0");
    dll_init(&m_socketcanobj.m_rcvList, SORT_NONE);
    dll_init(&m_socketcanobj.m_sndList, SORT_NONE);
}

CSocketcanObj::~CSocketcanObj()
{
    close(); // Close comm channel in case its open
    dll_removeAllNodes(&m_socketcanobj.m_rcvList);
    dll_removeAllNodes(&m_socketcanobj.m_sndList);
}


//------------------------------------------------------------------------------
// Open the CAN232 comm port
//
// Parameters are as follows
//
// interface;mask;filter
//
//------------------------------------------------------------------------------

bool CSocketcanObj::open(const char *pDevice, unsigned long flags)
{
	int rv = true;
	//char devname[IFNAMSIZ + 1];
	//fd_set rdfs;
	//struct timeval tv;
	//struct sockaddr_can addr;        
	//struct ifreq ifr;
    //struct canfd_frame frame;
	//struct cmsghdr *cmsg;	
	//char ctrlmsg[CMSG_SPACE(sizeof(struct timeval)) + CMSG_SPACE(sizeof(__u32))];
	//const int canfd_on = 1;
	
	// No device name
	memset(m_socketcanobj.m_devname, 0, sizeof(m_socketcanobj.m_devname) );
	
	//----------------------------------------------------------------------
	//	Set default parameters
	//----------------------------------------------------------------------
	strcpy(m_socketcanobj.m_devname, "vcan0");
	unsigned long nMask = 0;
	unsigned long nFilter = 0;


	//----------------------------------------------------------------------
	//	Acquire Mutex
	//----------------------------------------------------------------------
	pthread_attr_t thread_attr;
	pthread_attr_init(&thread_attr);
	pthread_mutex_init( &m_socketcanRcvMutex, NULL );
	pthread_mutex_init( &m_socketcanSndMutex, NULL );

	m_socketcanobj.m_bRun = true;

	//----------------------------------------------------------------------
	//	Parse given parameters
	//----------------------------------------------------------------------
	// Interface
	char *p = strtok((char *) pDevice, ";");
	if (NULL != p) {
		memset(m_socketcanobj.m_devname, 0, sizeof(m_socketcanobj.m_devname) );
		strncpy(m_socketcanobj.m_devname, p, strlen(p));
	}

	// Mask
	p = strtok(NULL, ";");
	if (NULL != p) {
		if ((NULL != strstr(p, "0x")) || (NULL != strstr(p, "0X"))) {
			sscanf(p + 2, "%lx", &nMask);
		} else {
			nMask = atol(p);
		}
	}

	// Filter
	p = strtok(NULL, ";");
	if (NULL != p) {
		if ((NULL != strstr(p, "0x")) || (NULL != strstr(p, "0X"))) {
			sscanf(p + 2, "%lx", &nFilter);
		} else {
			nFilter = atol(p);
		}
	}


	//----------------------------------------------------------------------
	//
	//----------------------------------------------------------------------

	// Initiate statistics
	m_socketcanobj.m_stat.cntReceiveData = 0;
	m_socketcanobj.m_stat.cntReceiveFrames = 0;
	m_socketcanobj.m_stat.cntTransmitData = 0;
	m_socketcanobj.m_stat.cntTransmitFrames = 0;

	m_socketcanobj.m_stat.cntBusOff = 0;
	m_socketcanobj.m_stat.cntBusWarnings = 0;
	m_socketcanobj.m_stat.cntOverruns = 0;

	//----------------------------------------------------------------------
	// Start thread 
	//----------------------------------------------------------------------
	if ( pthread_create( &m_threadId,
							&thread_attr,
							workThread,
							this)) {
		rv = false;
		close();
	}

	//----------------------------------------------------------------------
	// Release the mutex for other threads to use
	//----------------------------------------------------------------------
	pthread_mutex_unlock(&m_socketcanRcvMutex);
	pthread_mutex_unlock(&m_socketcanSndMutex);

	return rv;
}


//------------------------------------------------------------------------------
//  close
//------------------------------------------------------------------------------

int CSocketcanObj::close(void)
{
	int rv = 0;

	// Do nothing if already terminated
	if (!m_socketcanobj.m_bRun) return 1;

	// Terminate the thread
	m_socketcanobj.m_bRun = false;

	// Give the worker thread some time to terminate
	int *trv;
	pthread_join(m_threadId, (void **) &trv);
	pthread_mutex_destroy(&m_socketcanRcvMutex);
	pthread_mutex_destroy(&m_socketcanSndMutex);

	::close(m_socketcanobj.m_sock);

	return rv;
}

//------------------------------------------------------------------------------
//  writeMsg
//------------------------------------------------------------------------------

int CSocketcanObj::writeMsg(PCANALMSG pCanalMsg)
{
	int rv = 0;

	// Must be room for the message
	if (m_socketcanobj.m_sndList.nCount < SOCKETCAN_MAX_SNDMSG) {
		if (NULL != pCanalMsg) {
			dllnode *pNode = new dllnode;
			if (NULL != pNode) {
				canalMsg *pnewMsg = new canalMsg;
				pNode->pObject = pnewMsg;
				pNode->pKey = NULL;
				pNode->pstrKey = NULL;
				if (NULL != pnewMsg) {
					memcpy(pnewMsg, pCanalMsg, sizeof( canalMsg));
				}

				LOCK_MUTEX(m_socketcanSndMutex);
				dll_addNode(&m_socketcanobj.m_sndList, pNode);
				UNLOCK_MUTEX(m_socketcanSndMutex);

				rv = true;
			}
		}
	}
    
	return rv;
}

//------------------------------------------------------------------------------
//  readMsg
//------------------------------------------------------------------------------

int CSocketcanObj::readMsg(canalMsg *pMsg)
{
	int rv = false;

	if ((NULL != m_socketcanobj.m_rcvList.pHead) &&
			(NULL != m_socketcanobj.m_rcvList.pHead->pObject)) {

		memcpy(pMsg, m_socketcanobj.m_rcvList.pHead->pObject, sizeof( canalMsg));
		
		LOCK_MUTEX(m_socketcanRcvMutex);
		dll_removeNode(&m_socketcanobj.m_rcvList, m_socketcanobj.m_rcvList.pHead);
		UNLOCK_MUTEX(m_socketcanRcvMutex);

		rv = true;
	}
	return rv;
}

//------------------------------------------------------------------------------
//	setFilter
//------------------------------------------------------------------------------

bool CSocketcanObj::setFilter(unsigned long filter, unsigned long mask)
{
	return true;
}

//------------------------------------------------------------------------------
//	setFilter
//------------------------------------------------------------------------------

bool CSocketcanObj::setFilter(unsigned long filter)
{
	return true;
}

//------------------------------------------------------------------------------
//	setMask
//------------------------------------------------------------------------------

bool CSocketcanObj::setMask(unsigned long mask)
{
	return true;
}

//------------------------------------------------------------------------------
//	dataAvailable
//------------------------------------------------------------------------------

int CSocketcanObj::dataAvailable(void)
{
	int cnt;

	LOCK_MUTEX(m_socketcanRcvMutex);
	cnt = dll_getNodeCount(&m_socketcanobj.m_rcvList);
	UNLOCK_MUTEX(m_socketcanRcvMutex);

	return cnt;
}

//------------------------------------------------------------------------------
//	getStatistics
//------------------------------------------------------------------------------

bool CSocketcanObj::getStatistics(PCANALSTATISTICS& pCanalStatistics)
{
	pCanalStatistics = &m_socketcanobj.m_stat;
	return true;
}

//------------------------------------------------------------------------------
//	getStatus
//------------------------------------------------------------------------------

bool CSocketcanObj::getStatus(PCANALSTATUS pCanalStatus)
{
	return true;
}

//------------------------------------------------------------------------------
// workThread
//
// The workThread do most of the actual work such as send and receive.
//------------------------------------------------------------------------------

void *workThread(void *pObject)
{
	int rv = 0;
	int sock;
	fd_set rdfs;
	struct timeval tv;
	struct sockaddr_can addr;
	struct ifreq ifr;
	struct can_frame frame;


	CSocketcanObj * psocketcanobj = (CSocketcanObj *) pObject;
	if (NULL == psocketcanobj) {
		pthread_exit(&rv);
	}
	
	sock = socket(PF_CAN, SOCK_RAW, CAN_RAW);
	if (sock < 0) {
		syslog(LOG_ERR,
				"%s",
				(const char *) "CReadSocketCanTread: Error while opening socket. Terminating!");
		return NULL;
	}

	strcpy(ifr.ifr_name, psocketcanobj->m_socketcanobj.m_devname );
	ioctl(sock, SIOCGIFINDEX, &ifr);
	
	addr.can_family = AF_CAN;
	addr.can_ifindex = ifr.ifr_ifindex;

#ifdef DEBUG
	printf("using interface name '%s'.\n", ifr.ifr_name);
#endif

	
	if ( bind( sock, (struct sockaddr *)&addr, sizeof(addr) ) < 0 ) {
		syslog(LOG_ERR,
				"%s",
				(const char *) "CReadSocketCanTread: Error in socket bind. Terminating!");
		return NULL;
	}
	

	while (psocketcanobj->m_socketcanobj.m_bRun) {

		///////////////////////////////////////////////////////////////////////
		//                        Receive 
		///////////////////////////////////////////////////////////////////////

		//LOCK_MUTEX(psocketcanobj->m_socketcanObjMutex);

		// Noting to do if we should end...
		if (!psocketcanobj->m_socketcanobj.m_bRun) continue;

		FD_ZERO(&rdfs);
		FD_SET(sock, &rdfs);

		tv.tv_sec = 0;
		tv.tv_usec = 50; // 50ms timeout 5000

		int ret;
		if ((ret = select(sock+1, &rdfs, NULL, NULL, &tv)) < 0) {
			// Error
			psocketcanobj->m_socketcanobj.m_bRun = true;
			continue;
		}

		if (ret) {

			// There is data to read
			ret = read(sock, &frame, sizeof(struct can_frame));
			if (ret < 0) {
				psocketcanobj->m_socketcanobj.m_bRun = true;
				continue;
			}

			if (psocketcanobj->m_socketcanobj.m_rcvList.nCount < SOCKETCAN_MAX_RCVMSG) {
				
				PCANALMSG pMsg = new canalMsg;
				
				if (NULL != pMsg) {
                    pMsg->flags = 0;
					dllnode *pNode = new dllnode;
					if (NULL != pNode) {
							
						pNode->pObject = pMsg;
						
						// Assign the message
						pMsg->flags = 0;
						if (frame.can_id & CAN_RTR_FLAG) pMsg->flags |= CANAL_IDFLAG_RTR;
						if (frame.can_id & CAN_EFF_FLAG) {
							pMsg->id = frame.can_id & CAN_EFF_MASK;
							pMsg->flags |= CANAL_IDFLAG_EXTENDED;
						}
						else {
							pMsg->id = frame.can_id & CAN_SFF_MASK;
						}
						pMsg->sizeData = frame.can_dlc;
						memcpy(pMsg->data,frame.data,frame.can_dlc);
						
						LOCK_MUTEX(psocketcanobj->m_socketcanRcvMutex);
						dll_addNode(&psocketcanobj->m_socketcanobj.m_rcvList, pNode);
						UNLOCK_MUTEX(psocketcanobj->m_socketcanRcvMutex);
						
						// Update statistics
						psocketcanobj->m_socketcanobj.m_stat.cntReceiveData += pMsg->sizeData;
						psocketcanobj->m_socketcanobj.m_stat.cntReceiveFrames += 1;			
						
					} 
					else {
						delete pMsg;
					}
				}
			} // room in queue
						
			
		}

		//UNLOCK_MUTEX(psocketcanobj->m_socketcanObjMutex);

		///////////////////////////////////////////////////////////////////////
		//                          Transmit
		///////////////////////////////////////////////////////////////////////

		

		if ((NULL != psocketcanobj->m_socketcanobj.m_sndList.pHead) &&
				(NULL != psocketcanobj->m_socketcanobj.m_sndList.pHead->pObject)) {

			canalMsg msg;

			memcpy(&msg, psocketcanobj->m_socketcanobj.m_sndList.pHead->pObject, sizeof( canalMsg));
			
			LOCK_MUTEX(psocketcanobj->m_socketcanSndMutex);
			dll_removeNode(&psocketcanobj->m_socketcanobj.m_sndList, psocketcanobj->m_socketcanobj.m_sndList.pHead);			
			UNLOCK_MUTEX(psocketcanobj->m_socketcanSndMutex);
			
			memset(&frame, 0, sizeof(struct can_frame));
			frame.can_id = msg.id;
			frame.can_id |= CAN_EFF_FLAG; // Always extended
			frame.can_dlc = msg.sizeData;

			memcpy(frame.data, msg.data, msg.sizeData);
			
			// Write the data
			write( sock, &frame, sizeof(struct can_frame) );

			// Update statistics
			psocketcanobj->m_socketcanobj.m_stat.cntTransmitData += msg.sizeData;
			psocketcanobj->m_socketcanobj.m_stat.cntTransmitFrames += 1;

		} // if there is something to transmit

	} // while( psocketcanobj->m_socketcanobj.m_bRun )
	
	pthread_exit(&rv);
}
